zyte_common_items/pipelines.py

```diff
@@ -53,3 +55,67 @@ def __init__(self):
 
     def process_item(self, item, spider):
         return ae.downgrade(item)
+
+
+class DropLowProbabilityItemPipeline:
+    """This pipeline drops an item if its probability, defined in the settings,
+    is less than the specified threshold.
+
+    There are two ways that thresholds can be defined per item type using
+    the `ITEM_PROBABILITY_THRESHOLDS` setting:
+
+    With paths to Item classes:
+        ITEM_PROBABILITY_THRESHOLDS = {
+            "zyte_common_items.items.Article": 0.1,
+            "zyte_common_items.items.Product": 0.2
+        }
+
+    With Item classes:
+        from zyte_common_items import Article, Product
+        ITEM_PROBABILITY_THRESHOLDS = {
+            Article: 0.1,
+            Product: 0.2
+        }
+    """
+
+    def __init__(self, crawler):
+        self.stats = crawler.stats
+        self.thresholds = {}
+        self.init_thresholds(crawler.spider)
+
+    @classmethod
+    def from_crawler(cls, crawler):
+        return cls(crawler)
+
+    def init_thresholds(self, spider):
+        from scrapy.utils.misc import load_object
+
+        thresholds_settings = spider.settings.get("ITEM_PROBABILITY_THRESHOLDS", {})
+        for item, threshold in thresholds_settings.items():
+            item_type = load_object(item) if isinstance(item, str) else item
+            self.thresholds[item_type] = threshold
+
+    def get_threshold(self, item, spider):
+        return self.thresholds.get(type(item), DEFAULT_ITEM_PROBABILITY_THRESHOLD)
+
+    def get_item_name(self, item):
+        item_name = item.__class__.__name__
+        return item_name[0].lower() + item_name[1:]
+
+    def process_item(self, item, spider):
+        import scrapy
+
+        threshold = self.get_threshold(item, spider)
+
+        self.stats.inc_value("item/crawl/total", spider=spider)
+        item_proba = item.get_probability()
+        if item_proba is None or item_proba >= threshold:
+            return item
```

Comment:
Maybe we could do this to allow items with `None` as probability to be yielded if their threshold is 0?

```suggestion
        item_proba = item.get_probability() or 0
        if item_proba >= threshold:
            return item
```

Reply:
Should it be `or 1`?

Reply:
The current setup is inline with the ecommerce setup: https://github.com/zytedata/zyte-spider-templates/blob/main/zyte_spider_templates/spiders/ecommerce.py#L168

Reply:
Oh, I misunderstood the meaning of `None`, and I thought when `None` items were being filtered out, not returned :facepalm: